File: data/recomposed_projects/llama-3-70b-instruct/body/commons-codec/src/main/org/apache/commons/codec/language/bm/Languages.py
from __future__ import annotations
import re
from abc import ABC
import io
import typing
from typing import *
from src.main.org.apache.commons.codec.Resources import *
from src.main.org.apache.commons.codec.language.bm.NameType import *
from src.main.org.apache.commons.codec.language.bm.ResourceConstants import *


class LanguageSet(ABC):

    @staticmethod
    def from_(langs: typing.Set[str]) -> LanguageSet:
        return NO_LANGUAGES if langs.isEmpty() else SomeLanguages(langs)

    def merge(self, other: LanguageSet) -> LanguageSet:
        return LanguageSet()

    def restrictTo(self, other: LanguageSet) -> LanguageSet:
        return LanguageSetIntersection(self, other)

    def isSingleton(self) -> bool:
        return self.size() == 1

    def isEmpty(self) -> bool:
        return self.languages == None or self.languages.isEmpty()

    def getAny(self) -> str:
        return self.getLanguages().get(0)

    def contains(self, language: str) -> bool:
        return language in self.languages


class AnyLanguage(LanguageSet):

    def toString(self) -> str:
        return "ANY_LANGUAGE"

    def merge(self, other: LanguageSet) -> LanguageSet:
        return other

    def restrictTo(self, other: LanguageSet) -> LanguageSet:
        return other

    def isSingleton(self) -> bool:
        return False

    def isEmpty(self) -> bool:
        return False

    def getAny(self) -> str:
        raise RuntimeError("Can't fetch any language from the any language set.")

    def contains(self, language: str) -> bool:
        return True

    def __init__(self) -> None:
        super().__init__()


class NoLanguage(LanguageSet):

    def toString(self) -> str:
        return "NO_LANGUAGES"

    def merge(self, other: LanguageSet) -> LanguageSet:
        return other

    def restrictTo(self, other: LanguageSet) -> LanguageSet:
        return self

    def isSingleton(self) -> bool:
        return False

    def isEmpty(self) -> bool:
        return True

    def getAny(self) -> str:
        raise RuntimeError("Can't fetch any language from the empty language set.")

    def contains(self, language: str) -> bool:
        return False

    def __init__(self) -> None:
        super().__init__()


class SomeLanguages(LanguageSet):

    __languages: typing.Set[str] = None

    def toString(self) -> str:
        return "Languages(" + self.__languages.toString() + ")"

    def merge(self, other: LanguageSet) -> LanguageSet:
        if other == NO_LANGUAGES:
            return self
        if other == ANY_LANGUAGE:
            return other
        some_languages = typing.cast(SomeLanguages, other)
        set_ = set(self.__languages)
        set_.update(some_languages.__languages)
        return LanguageSet.from_(set_)

    def restrictTo(self, other: LanguageSet) -> LanguageSet:
        if other == NO_LANGUAGES:
            return other
        if other == ANY_LANGUAGE:
            return self
        someLanguages = typing.cast(SomeLanguages, other)
        set = set()
        for lang in self.__languages:
            if lang in someLanguages.__languages:
                set.add(lang)
        return LanguageSet.from_(set)

    def isSingleton(self) -> bool:
        return len(self.__languages) == 1

    def isEmpty(self) -> bool:
        return self.__languages.isEmpty()

    def getAny(self) -> str:
        return next(iter(self.__languages))

    def contains(self, language: str) -> bool:
        return self.__languages.contains(language)

    def getLanguages(self) -> typing.Set[str]:
        return self.__languages

    def __init__(self, languages: typing.Set[str]) -> None:
        self.__languages = frozenset(languages)


class Languages:

    ANY_LANGUAGE: LanguageSet = None
    ANY_LANGUAGE: LanguageSet = None
    NO_LANGUAGES: LanguageSet = None
    NO_LANGUAGES: LanguageSet = None
    ANY: str = "any"
    __languages: typing.Set[str] = None

    __LANGUAGES: typing.Dict[NameType, Languages] = {}

    @staticmethod
    def initialize_fields() -> None:
        Languages.ANY_LANGUAGE: LanguageSet = AnyLanguage()

        Languages.ANY_LANGUAGE: LanguageSet = AnyLanguage()

        Languages.NO_LANGUAGES: LanguageSet = NoLanguage()

        Languages.NO_LANGUAGES: LanguageSet = NoLanguage()

        Languages.__LANGUAGES: typing.Dict[NameType, Languages] = {}

    @staticmethod
    def run_static_init():
        for s in NameType:
            Languages.__LANGUAGES[s] = Languages.getInstance1(
                Languages.__langResourceName(s)
            )

    def getLanguages(self) -> typing.Set[str]:
        return self.__languages

    @staticmethod
    def getInstance1(languagesResourceName: str) -> Languages:
        ls: typing.Set[str] = set()
        with Resources.getInputStream(languagesResourceName) as lsScanner:
            inExtendedComment: bool = False
            while lsScanner.hasNextLine():
                line: str = lsScanner.nextLine().trim()
                if inExtendedComment:
                    if line.endsWith(ResourceConstants.EXT_CMT_END):
                        inExtendedComment = False
                else:
                    if line.startsWith(ResourceConstants.EXT_CMT_START):
                        inExtendedComment = True
                    else:
                        if not line.isEmpty():
                            ls.add(line)
        return Languages(Collections.unmodifiableSet(ls))

    @staticmethod
    def getInstance0(nameType: NameType) -> Languages:
        return Languages.__LANGUAGES.get(nameType)

    def __init__(self, languages: typing.Set[str]) -> None:
        self.__languages = languages

    @staticmethod
    def __langResourceName(nameType: NameType) -> str:
        return (
            f"org/apache/commons/codec/language/bm/{nameType.getName()}_languages.txt"
        )


Languages.initialize_fields()

Languages.run_static_init()

File: java_projects/cleaned_final_projects_evosuite/commons-codec/src/main/java/org/apache/commons/codec/language/bm/Languages.java

/*
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */


package org.apache.commons.codec.language.bm;

import org.apache.commons.codec.Resources;

import java.util.Collections;
import java.util.EnumMap;
import java.util.HashSet;
import java.util.Map;
import java.util.NoSuchElementException;
import java.util.Scanner;
import java.util.Set;

/**
 * Language codes.
 *
 * <p>Language codes are typically loaded from resource files. These are UTF-8 encoded text files.
 * They are systematically named following the pattern:
 *
 * <blockquote>
 *
 * org/apache/commons/codec/language/bm/${{@link NameType#getName()} languages.txt
 *
 * </blockquote>
 *
 * <p>The format of these resources is the following:
 *
 * <ul>
 *   <li><b>Language:</b> a single string containing no whitespace
 *   <li><b>End-of-line comments:</b> Any occurrence of '//' will cause all text following on that
 *       line to be discarded as a comment.
 *   <li><b>Multi-line comments:</b> Any line starting with '/*' will start multi-line commenting
 *       mode. This will skip all content until a line ending in '*' and '/' is found.
 *   <li><b>Blank lines:</b> All blank lines will be skipped.
 * </ul>
 *
 * <p>Ported from language.php
 *
 * <p>This class is immutable and thread-safe.
 *
 * @since 1.6
 */
public class Languages {

    /** A set of languages. */
    public abstract static class LanguageSet {

        public static LanguageSet from(final Set<String> langs) {
            return langs.isEmpty() ? NO_LANGUAGES : new SomeLanguages(langs);
        }

        public abstract boolean contains(String language);

        public abstract String getAny();

        public abstract boolean isEmpty();

        public abstract boolean isSingleton();

        public abstract LanguageSet restrictTo(LanguageSet other);

        abstract LanguageSet merge(LanguageSet other);
    }

    public static class AnyLanguage extends LanguageSet {
        public AnyLanguage() {
            super();
        }
    
        public boolean contains(final String language) {
            return true;
        }

        public String getAny() {
            throw new NoSuchElementException(
                    "Can't fetch any language from the any language set.");
        }

        public boolean isEmpty() {
            return false;
        }

        public boolean isSingleton() {
            return false;
        }

        public LanguageSet restrictTo(final LanguageSet other) {
            return other;
        }

        public LanguageSet merge(final LanguageSet other) {
            return other;
        }

        public String toString() {
            return "ANY_LANGUAGE";
        }
    }

    public static class NoLanguage extends LanguageSet {
        public NoLanguage() {
            super();
        }
        
        public boolean contains(final String language) {
            return false;
        }

        public String getAny() {
            throw new NoSuchElementException(
                    "Can't fetch any language from the empty language set.");
        }

        public boolean isEmpty() {
            return true;
        }

        public boolean isSingleton() {
            return false;
        }

        public LanguageSet restrictTo(final LanguageSet other) {
            return this;
        }

        public LanguageSet merge(final LanguageSet other) {
            return other;
        }

        public String toString() {
            return "NO_LANGUAGES";
        }
    }

    /** Some languages, explicitly enumerated. */
    public static final class SomeLanguages extends LanguageSet {
        private final Set<String> languages;

        private SomeLanguages(final Set<String> languages) {
            this.languages = Collections.unmodifiableSet(languages);
        }

        @Override
        public boolean contains(final String language) {
            return this.languages.contains(language);
        }

        @Override
        public String getAny() {
            return this.languages.iterator().next();
        }

        public Set<String> getLanguages() {
            return this.languages;
        }

        @Override
        public boolean isEmpty() {
            return this.languages.isEmpty();
        }

        @Override
        public boolean isSingleton() {
            return this.languages.size() == 1;
        }

        @Override
        public LanguageSet restrictTo(final LanguageSet other) {
            if (other == NO_LANGUAGES) {
                return other;
            }
            if (other == ANY_LANGUAGE) {
                return this;
            }
            final SomeLanguages someLanguages = (SomeLanguages) other;
            final Set<String> set =
                    new HashSet<>(Math.min(languages.size(), someLanguages.languages.size()));
            for (final String lang : languages) {
                if (someLanguages.languages.contains(lang)) {
                    set.add(lang);
                }
            }
            return from(set);
        }

        @Override
        public LanguageSet merge(final LanguageSet other) {
            if (other == NO_LANGUAGES) {
                return this;
            }
            if (other == ANY_LANGUAGE) {
                return other;
            }
            final SomeLanguages someLanguages = (SomeLanguages) other;
            final Set<String> set = new HashSet<>(languages);
            set.addAll(someLanguages.languages);
            return from(set);
        }

        @Override
        public String toString() {
            return "Languages(" + languages.toString() + ")";
        }
    }

    public static final String ANY = "any";

    private static final Map<NameType, Languages> LANGUAGES = new EnumMap<>(NameType.class);

    static {
        for (final NameType s : NameType.values()) {
            LANGUAGES.put(s, getInstance1(langResourceName(s)));
        }
    }

    public static Languages getInstance0(final NameType nameType) {
        return LANGUAGES.get(nameType);
    }

    public static Languages getInstance1(final String languagesResourceName) {
        final Set<String> ls = new HashSet<>();
        try (final Scanner lsScanner =
                new Scanner(
                        Resources.getInputStream(languagesResourceName),
                        ResourceConstants.ENCODING)) {
            boolean inExtendedComment = false;
            while (lsScanner.hasNextLine()) {
                final String line = lsScanner.nextLine().trim();
                if (inExtendedComment) {
                    if (line.endsWith(ResourceConstants.EXT_CMT_END)) {
                        inExtendedComment = false;
                    }
                } else if (line.startsWith(ResourceConstants.EXT_CMT_START)) {
                    inExtendedComment = true;
                } else if (!line.isEmpty()) {
                    ls.add(line);
                }
            }
            return new Languages(Collections.unmodifiableSet(ls));
        }
    }

    private static String langResourceName(final NameType nameType) {
        return String.format(
                "org/apache/commons/codec/language/bm/%s_languages.txt", nameType.getName());
    }

    private final Set<String> languages;

    /** No languages at all. */
    public static final LanguageSet NO_LANGUAGES = new NoLanguage();

    /** Any/all languages. */
    public static final LanguageSet ANY_LANGUAGE = new AnyLanguage();

    private Languages(final Set<String> languages) {
        this.languages = languages;
    }

    public Set<String> getLanguages() {
        return this.languages;
    }
}
